Add bookings

Nucleus1337 opened the pull request.

Nucleus1337 committed:
feat: добавить зависимость

Nucleus1337 committed:
feat: init commit

Nucleus1337 committed:
fix: добавить if exists при удалении таблиц

SemenSH commented on src/main/java/ru/practicum/shareit/booking/Booking.java:
```diff
@@ -1,19 +1,51 @@
 import ru.practicum.shareit.user.User;
 
 /** TODO Sprint add-bookings. */
 @Data
```
данную аннотацию лучше не использовать для jpa моделей, equals/hashcode переопределить вручную c одним гарантированно уникальным полем - id. Применить для всех моделей

Nucleus1337 replied:
Исправил.

SemenSH commented on src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -1,12 +1,62 @@
+  @PostMapping
+  @Validated({Group.OnInsert.class})
+  public BookingResponseDto create(
+      @RequestHeader("X-Sharer-User-Id") Long userId,
```
значение хэдэров лучше вынести в константу

Nucleus1337 replied:
Исправил.

SemenSH commented on src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -1,12 +1,62 @@
+  public BookingResponseDto create(
+      @RequestHeader("X-Sharer-User-Id") Long userId,
+      @RequestBody @Valid BookingRequestDto bookingRequestDto) {
+    return bookingService.create(userId, bookingRequestDto);
```
для каждого эндпойнта в контроллерах нужно добавить логирование, т.к. это важные точки приложения и при отладке нужно понимать, прошел запрос или нет и с какими данными

Nucleus1337 replied:
Сделал.

SemenSH commented on src/main/java/ru/practicum/shareit/booking/BookingRepository.java:
```diff
@@ -0,0 +1,55 @@
+          + "and (i.owner.id = ?2 or b.booker.id = ?2)")
+  Optional<Booking> findBookingByIdAndOwnerIdOrBookerId(Long bookingId, Long userId);
+
+  @Query("select b from Booking b join fetch b.item i where i.owner.id = ?1 order by b.start desc")
```
большинство запросов можно реализовать с помощью Spring Data по наименованию метода

Nucleus1337 replied:
Решил оставить, как есть. Запросы понимать мне проще, чем длинные названия методов, если это не влияет на быстродействие.

SemenSH commented on src/main/java/ru/practicum/shareit/booking/BookingService.java:
```diff
@@ -0,0 +1,143 @@
+        .orElseThrow(() -> new CustomException.UserNotFoundException("Пользователь не найден"));
+  }
+
+  /** TODO: можно написать свой валидатор; оставил на будущее */
```
лучшим вариантом было бы создание своей аннотации и валидатора

Nucleus1337 replied:
Сделал.

SemenSH reviewed (commented):
Привет! Хорошая работа, есть небольшие замечания, по вопросам ревью в пачку

Nucleus1337 committed:
fix: @Data -> @Getter + @Setter + equal + hashcode;
вынести значения хедеров в контролерах в константу;
feat: добавить валидатор для дат бронирования;
добавить логирования для контроллеров;

Nucleus1337 committed:
fix: поправить application.properties для тестов

SemenSH reviewed (commented):
Замечаний нет. Успехов в следующем спринте
